### ISSUE TITLE
Evaluating @arguments

### ISSUE THREAD

**Issue by Justineo** [less/less.js#2436]
Hi guys.

For the following Less code:

```less
.mixin(@param) {
  @x: @arguments;

  .output(@whatever) {
    x: @x;
  }

  .output(unexpected);
}

.test {
  .mixin(expected);
}
```

I would expect the rendered CSS to be like this:

```css
.test {
  x: expected;
}
```

But actually it is:

```css
.test {
  x: unexpected;
}
```

I know that Less lazy loads variables but this still seems weird to me. Shouldn't things like `@arguments` and `@rest` be evaluated before it goes into deeper scope?

**Comment by seven-phases-max** [comment]
I added "Bug" label at first, but I'm quite confused now. Notice that it not so different from the second example in [Lazy Loading](http://lesscss.org/features/#variables-feature-lazy-loading).  Also it's not limited to only `@arguments` variable and/or `~""` stuff. It' s for any variable inside `.output`:
```
.mixin(@param) {
  @x: @arguments @param;
  .output(@param) {
    x: @x;
  }

  .output(unexpected);
}

.test {
  .mixin(expected); // unexpected unexpected
}
```
Just in case I also tested earlier versions (down to 1.4.2) and it's always been like this.

**Comment by Justineo** [comment]
Yes it seems to be consistent with lazy loading but still weird...
In common cases when we use `@arguments`, we are actually expecting it will be evaluated within current scope. In my actual code I can refactor a little bit to avoid this but the logic here is quite confusing.

**Comment by Justineo** [comment]
@seven-phases-max Like you said it's not limited to `@arguments`. But in other cases I can avoid the unexpected behavior by naming parameters in deeper scopes different from variables from upper scopes. `@arguments` is automatically generated in each scope and so this kind of workaround won't work.

**Comment by seven-phases-max** [comment]
@Justineo

Since I mentioned this issue in https://github.com/less/less.js/issues/2433#issuecomment-73204737,
I feel like I need to provide an alternative answer now.

Regardless of how we'll decide to treat that (change or not) the root of the problem in the example is of course the use of the caller variable in the called mixin, in Less it's a perfect way to shoot yourself in the foot. Such case always should be taken with care and avoided when possible (the problem is that there're too many directly contradictory scope handling expectations and already implemented rules so it will always have "issues").

So yes, even if we decide to change it for `@arguments` it's still better to refactor your code to pass caller variables to the mixin explicitly (just to feel yourself safer :), e.g.:  
```

.mixin(@param) {
    .output(@x) {
        x: @x;
    }

    .output(@arguments);
}

.test {
    .mixin(expected);
}
```

**Comment by rjgotten** [comment]
More confirmation for what has been on my mind for some time now:
LESS's attempts at being clever with lazy evaluation and its often nonsensical and non-orthogonal/non-uniform scoping rules are **not a good thing**.

Of late there have been far too many instances of people shooting themselves in the foot or the compiler containing actual bugs, all in major part due to trying to cope with the quirks and exceptions in these rules. Perhaps it's time to make work of missing crucial features such as lambdas and function return values and just switch to passing by parameter, closure scope or return value *only*: neat, nice and above all *predictable*.

As for a migration path and backwards compatibility: we already have `strict-units` and `strict-math`. There is no reason we cannot have `strict-scope` as well.

**Comment by seven-phases-max** [comment]
@rjgotten  

Well, strictly speaking, Less scoping rules are uniform and strict. Even the most famous #1316 (that applies here as well) *is not* caused by actually *different* (or some "clever") scoping rules but is only a result of different "use" of plain rulesets and parametric mixins (I explained it in details [here]).
The only thing that may be considered "clever" in this context is "variabls expanded by mixin do not override existing variables" but it's pretty much issue-free feature (and it's not like "it works this way in this case and another way in that case").

>Perhaps it's time to make work of missing crucial features such as lambdas and function return values

Ready to make a PR for all this? ;) OK, just killing. But in this context I'd like to remind that initially Less is meant to be not *programming* language at all (in classical sense of this term, just like CSS isn't). And while recently it's got some of such functionality, expecting all these not-even-considered-initially features (callbacks, lamdas, functions, [OOP classes]) to appear instantly and fit like a charm out-of-the-box into existing facilities is rather strange.

**Comment by rjgotten** [comment]
>  an approach of "implement what I need, redesign what I don't like, throw away what I don't need, provide an option to disable all that so it won't break existing code and it's all solved" just won't work. 

Well; lambdas and real mixin return values are general additions that don't change existing behavior. The scoping rules are the only real change: as far as I've been able to tell, mostly everything is already using normal closure scope, but pulls in stuff from other scopes *in addition* to that. A `strict-scope` switch would only turn *that* behavior off and nothing else.

> Ready to make a PR for all this? ;) 

Sorry; kind of busy with `@import (plugin)` already. ;-)

**Comment by seven-phases-max** [comment]
Yes, sure, lambdas and mixin return values won't be affected by and/or won't change scoping rules, hence `strict-scope` option won't make any sense for them anyway...  So my comment is mostly for those feature-request/issues that scope-priority related issues.

>A `strict-scope` switch would only turn that behavior off and nothing else.

Just turning caller scope off? OK, makes sense then... to solve this one (Sorry, I was thinking you also mean other scope issues).

**Comment by seven-phases-max** [comment]
After some thinking Im about closing this (feel free to reopen).
As already mentioned the result is expected because of lazy-evaluation (aka lazy-loading), the value of `@arguments` is evaluated when `x: @x` is rendered (and not when `@arguments` is assigned to `@x`) hense it fall to `..output` arguments. *No* caller/parent precedence quirks are actually involved here.

And while it would seem to be natural to make *an exception* for `@arguments` because of its special meaning, in general this would mean actually introducing that uncertain and inconsistent evaluation hack also mentioned above and bringing just more chaos in.
In other words "lazy-evaluation" is "lazy-evaluation". *No* exceptions.

---
For the particular example of the initial post this means it should not rely on the special `@arguments` meaning, but use explicit variable instead, e.g.:
```less
.mixin(@parameters...) {
    @x: @parameters;
    
    .output(@whatever) {
        x: @x;
    }

    .output(unexpected);
}

.test {
    .mixin(expected);
}
```

**Comment by seven-phases-max** [comment]
Closing as expected behaviour.
